_posts/2019-12-20-differential-privacy/differential_privacy.Rmd

```diff
@@ -0,0 +1,470 @@
+knitr::include_graphics("images/baseline.png")
+```
+
+Just to put this in context, the MAE reported for subject S1 in the paper[@ReissISL19] -- based on a higher-capacity network, extensive hyperparameter tuning, and naturally, training on the complete dataset -- amounts to 8.45 bpm on average; so our setup seems to be sound.
+
+Now we'll make this differentially private.
+
+
+## DP training
+
+Instead of plain `Adam`, we use the corresponding TF Privacy wrapper, `DPAdamGaussianOptimizer`.
+
+We need to tell it how aggressive gradient clipping should be (`l2_norm_clip`) and how much noise to add (`noise_multiplier`). Furthermore, we define the learning rate (there is no default), going for 10 times the default `0.001` based on initial experiments.
+
+There is an additional parameter, `num_microbatches`, that could be used to speed up training [@abs181206210], but as training duration is not an issue here we just set it equal to `batch_size`.
+
+The values for `l2_norm_clip` and `noise_multiplier` chosen here follow those used in the [tutorials in the TF Privacy repo](https://github.com/tensorflow/privacy/tree/master/tutorials).
+
+Nicely, TF Privacy comes with a script that allows one to compute the attained $\epsilon$ beforehand, based on number of training examples, `batch_size`, `noise_multiplier` and number of training epochs.^[There is an additional parameter, $\delta$, that allows for bounding the risk that the privacy guarantee does not hold. The [recommendation](https://github.com/tensorflow/privacy/tree/master/tutorials) is to set this to at most the inverse of the number of training examples, which in out case would mean <= ~ `1e-04`; the default setting is `1e-06` so we should be fine here.]
+
+Calling that [script](https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/analysis/compute_dp_sgd_privacy.py), and assuming we train for 20 epochs here as well, 
+
+```{python}
+python compute_dp_sgd_privacy.py --N=3989 --batch_size=32 --noise_multiplier=1.1 --epochs=20
+```
+
+this is what we get back:
+
+```
+DP-SGD with sampling rate = 0.802% and noise_multiplier = 1.1 iterated over
+2494 steps satisfies differential privacy with eps = 2.73 and delta = 1e-06.
+```
+
+How good is a value of 2.73? Citing the [TF Privacy authors](https://github.com/tensorflow/privacy/tree/master/tutorials):
+
+> $\epsilon$ gives a ceiling on how much the probability of a particular output can increase by including (or removing) a single training example. We usually want it to be a small constant (less than 10, or, for more stringent privacy guarantees, less than 1). However, this is only an upper bound, and a large value of epsilon may still mean good practical privacy.
+
+
+Obviously, choice of $\epsilon$ is a (challenging) topic by itself, and not something we can elaborate on in a post dedicated to the technical aspects of DP with TensorFlow.
```

Comment:
```suggestion
Obviously, choice of $\epsilon$ is a (challenging) topic unto itself, and not something we can elaborate on in a post dedicated to the technical aspects of DP with TensorFlow.
```

Reply:
"unto itself" ... need to keep that in mind (wasn't aware of)